### PR TITLE
refactor: making TaskManager responsible of creating its entry in StatusBarRegistry

### PR DESCRIPTION
**Pull request by axel7083**
### What does this PR do?

Related to https://github.com/containers/podman-desktop/issues/4243 (part 1)

> See https://github.com/containers/podman-desktop/pull/4244#pullrequestreview-1843650350 for details

### CODE REVIEW

**Review by benoitf — COMMENTED** [review]
would you include the command in this PR or another one

as TaskManager should be responsible as well for the command that it's registering in the status bar

'show-task-manager'
```

    commandRegistry.registerCommand('show-task-manager', () => {
      apiSender.send('toggle-task-manager', '');
    });

```

**Review by benoitf — COMMENTED** [review]
need a test checking we create a status bar entry


also based on some other patterns, probably add the setEntry in a init() method rather than in the constructor

it's easier for mocking/testing

**Comment by axel7083** [comment]
> as TaskManager should be responsible as well for the command that it's registering in the status bar

This would create a circular dependency

`CommandRegistry` requires `Telemetry`
`Telemetry `requires `ConfigurationRegistry`
`ConfigurationRegistry` requires `NotificationRegistry`
`NotificationRegistry` requires `TaskManager`

Therefore, `TaskManager` cannot requires `CommandRegistry`

**Comment by benoitf** [comment]
> > as TaskManager should be responsible as well for the command that it's registering in the status bar
> 
> This would create a circular dependency
> 
> `CommandRegistry` requires `Telemetry` `Telemetry `requires `ConfigurationRegistry` `ConfigurationRegistry` requires `NotificationRegistry` `NotificationRegistry` requires `TaskManager`
> 
> Therefore, `TaskManager` cannot requires `CommandRegistry`

ok we need to break the dependency ConfigurationRegistry -> NotificationRegistry

**Comment by axel7083** [comment]
> ok we need to break the dependency ConfigurationRegistry -> NotificationRegistry

How ? We have a powerful direct usage 
https://github.com/containers/podman-desktop/blob/1b370f67f13a5938e025644a35e1fea0e10c1205/packages/main/src/plugin/configuration-registry.ts#L161-L168

**Comment by benoitf** [comment]
one proposal could be: (I'm sure there are many others)

- Create an empty array of notifications in index.ts
- we call init method of ConfigurationRegistry that returns a notification (or none) and we add that to the array
- when we create NotificationRegistry, we give all existing notifications to the object (and clear the array / make it invalid, whatever for subsequent push/addcalls)

**Comment by axel7083** [comment]
> one proposal could be: (I'm sure there are many others)
> 
> * Create an empty array of notifications in index.ts
> * we call init method of ConfigurationRegistry that returns a notification (or none) and we add that to the array
> * when we create NotificationRegistry, we give all existing notifications to the object (and clear the array / make it invalid, whatever for subsequent push/addcalls)

I feel this fall out of the scope of this PR

**Comment by axel7083** [comment]
> one proposal could be: (I'm sure there are many others)
> 
> * Create an empty array of notifications in index.ts
> * we call init method of ConfigurationRegistry that returns a notification (or none) and we add that to the array
> * when we create NotificationRegistry, we give all existing notifications to the object (and clear the array / make it invalid, whatever for subsequent push/addcalls)

rebased with dependency broken

File: packages/main/src/plugin/index.ts [hunk 1]
```diff
@@ -413,6 +411,9 @@ export class PluginSystem {
     const exec = new Exec(proxy);
 
     const commandRegistry = new CommandRegistry(apiSender, telemetry);
+    const taskManager = new TaskManager(apiSender, statusBarRegistry, commandRegistry);
+
```
Comment (benoitf) [inline packages/main/src/plugin/index.ts:416]: here we should call the init method

File: packages/main/src/plugin/task-manager.ts [hunk 2]
```diff
@@ -28,7 +30,32 @@ export class TaskManager {
 
   private tasks = new Map<string, Task>();
 
-  constructor(private apiSender: ApiSenderType) {}
+  constructor(
+    private apiSender: ApiSenderType,
+    private statusBarRegistry: StatusBarRegistry,
+    private commandRegistry: CommandRegistry,
+  ) {
+    this.init();
```
Comment (benoitf) [inline packages/main/src/plugin/task-manager.ts]: constructor should not call methods to match existing patterns

File: packages/main/src/plugin/task-manager.spec.ts [hunk 3]
```diff
@@ -164,7 +174,7 @@ test('return true if notificationTask', async () => {
 });
 
 test('return false if it is not a notificationTask', async () => {
-  const taskManager = new TaskManager(apiSender);
+  const taskManager = new TaskManager(apiSender, statusBarRegistry, commandRegistry);
   const task = taskManager.createTask('title');
   const result = taskManager.isNotificationTask(task);
   expect(result).toBeFalsy();
```
Comment (benoitf) [inline packages/main/src/plugin/task-manager.spec.ts:185]: missing tests that init method is registering a command and registering a status bar entry

File: packages/main/src/plugin/task-manager.ts [hunk 4]
```diff
@@ -19,6 +19,8 @@
 import type { ApiSenderType } from './api.js';
 import type { NotificationInfo } from './api/notification.js';
 import type { NotificationTask, StatefulTask, Task } from './api/task.js';
+import type { StatusBarRegistry } from './statusbar/statusbar-registry.js';
```
Comment (benoitf) [inline packages/main/src/plugin/task-manager.ts:22]: here you didn't use `/@/` import but for the next line yes so it's looking odd

**Review by benoitf — COMMENTED** [review]
did a review yesterday but forgot to hit the 'submit' button...

File: packages/main/src/plugin/index.ts (hunk 1 above)
Comment (axel7083) [reply packages/main/src/plugin/index.ts:416]: Done

File: packages/main/src/plugin/task-manager.ts (hunk 2 above)
Comment (axel7083) [reply packages/main/src/plugin/task-manager.ts]: Agreed

File: packages/main/src/plugin/task-manager.ts (hunk 4 above)
Comment (axel7083) [reply packages/main/src/plugin/task-manager.ts:22]: My IDE is a crazy guy, I removed it

File: packages/main/src/plugin/task-manager.spec.ts (hunk 3 above)
Comment (axel7083) [reply packages/main/src/plugin/task-manager.spec.ts:185]: Added

File: packages/main/src/plugin/task-manager.ts [hunk 5]
```diff
@@ -28,7 +30,30 @@ export class TaskManager {
...
-  constructor(private apiSender: ApiSenderType) {}
+  constructor(
+    private apiSender: ApiSenderType,
+    private statusBarRegistry: StatusBarRegistry,
+    private commandRegistry: CommandRegistry,
+  ) {}
+
+  public init() {
+    // The TaskManager is responsible for creating the entry he will be using
+    this.statusBarRegistry.setEntry(
```
Comment (benoitf) [inline packages/main/src/plugin/task-manager.ts:41]: not for this PR but probably would be a nice enhancement on statusBarRegistry to return the Entry when calling setEntry method

**Review by benoitf — APPROVED** [review]

**Pull request merged**